packages/flutter/lib/src/widgets/editable.dart

```diff
@@ -43,11 +39,29 @@ class TextRange {
 
   /// Whether this range is empty (but still potentially placed inside the text).
   bool get isCollapsed => start == end;
+
+  /// The text before this range.
+  String textBefore(String text) {
+    return text.substring(0, start);
```

Comment:
That is an error condition.  I guess we could return null or truncate, but it seems reasonable to have the same error handling as `substring`.